Add embedding mode with arg flag. Currently working

StrikingLoo opened the pull request:
Hi everyone,
I took a stab at adding embedding mode, where we print the sentence embedding for the input instead of generating more tokens.
If I only add the compute and print in llama_eval itself, that works. But for some reason after adding the boolean flag (even if I create a second function that has the same preamble but in the end only prints the embeddings) it stops working.
Could anyone take a look and tell me where I failed or what I am missing, so that we can add this capability?

Thank you!

StrikingLoo committed:
working but ugly

StrikingLoo committed:
add arg flag, not working on embedding mode

StrikingLoo committed:
Merge branch 'master' of https://github.com/StrikingLoo/llama.cpp

StrikingLoo committed:
typo

nullhook commented on main.cpp:
```diff
@@ -936,12 +943,27 @@ int main(int argc, char ** argv) {
+        if (params.use_color) {
+            printf(ANSI_COLOR_RESET);
+        }
+        return 0;
```
Looks like you're returning without freeing.

`ggml_free(model.ctx);`

StrikingLoo commented:
It's true that I am, but we never reach that line. I forgot to uncomment it
for the PR, but before that, the program fails on the `ggml_graph_compute
(ctx0, &gf);` line.

We never reach that other part.

On Sun, Mar 19, 2023 at 8:09 PM taher ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In main.cpp
> <https://github.com/ggerganov/llama.cpp/pull/282#discussion_r1141579379>:
>
> > @@ -936,12 +943,27 @@ int main(int argc, char ** argv) {
>          printf(ANSI_COLOR_YELLOW);
>      }
>
> +    if (params.embedding){
> +        printf("got right before second call.\n");
> +        const int64_t t_start_us = ggml_time_us(); //HERE
> +        if (!llama_eval(model, params.n_threads, n_past, embd, logits, mem_per_token, true)) {
> +                fprintf(stderr, "Failed to predict\n");
> +                return 1;
> +        }
> +        //ggml_free(model.ctx);
> +
> +        if (params.use_color) {
> +            printf(ANSI_COLOR_RESET);
> +        }
> +        return 0;
>
> Looks like you're returning without freeing.
>
> ggml_free(model.ctx);
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ggerganov/llama.cpp/pull/282#pullrequestreview-1347726107>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD6JCX7LPLXDZHBKKHKKCHTW47C5XANCNFSM6AAAAAAV75U6WA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

nullhook commented on main.cpp:
```diff
@@ -936,12 +943,27 @@ int main(int argc, char ** argv) {
+    if (params.embedding){
+        printf("got right before second call.\n");
+        const int64_t t_start_us = ggml_time_us(); //HERE
+        if (!llama_eval(model, params.n_threads, n_past, embd, logits, mem_per_token, true)) {
```
You're not calling `llama_eval` for remaining tokens and doing other checks.

StrikingLoo replied:
the sentence embedding depends on the input, we shouldn't have to generate extra tokens later. Though I may be wrong. But changing that wouldn't solve the failed execution error.

nullhook replied:
Are you trying to capture input text's embeddings or predicted text embeddings?

adriacabeza replied:
Given the PR it is doing the input text's embeddings

nullhook replied:
The vector of tokens that you are sending for evaluation is empty. Please ensure that it is set correctly right before evaluating it, and also add a check to verify that it is not empty.

```c++
embd = embd_inp;

if (embd.size() > 0) {
   if (!llama_eval(model, params.n_threads, n_past, embd, logits, mem_per_token, true)) {
      fprintf(stderr, "Failed to predict\n");
      return 1;
   }
}
```

nullhook replied:
Let me know if the above fixes the runtime crash

StrikingLoo replied:
That solved it! My version is working but I see merge conflicts, I am testing the merge now

StrikingLoo replied:
Merge successful. Now we have embeddings! This should be ready to merge into main.

StrikingLoo commented on main.cpp:
```diff
@@ -936,12 +943,27 @@ int main(int argc, char ** argv) {
             const int64_t t_start_us = ggml_time_us();
 
-            if (!llama_eval(model, params.n_threads, n_past, embd, logits, mem_per_token)) {
+            if (!llama_eval(model, params.n_threads, n_past, embd, logits, mem_per_token, false)) {
```
But if we want the embedding, we only want to print once, and not generate the next token

nullhook commented:
Do the input tokens need to go through many layers to obtain their embeddings? My guess is that the input vectors should be obtainable earlier, so that it is not necessary to loop through multiple layers to obtain the final logits. Since logits are the final output, we should not process them when obtaining the embeddings.

StrikingLoo commented:
The embeddings should be the output of the last attention layer, corresponding to the last token in the input.
Say I input "I am a dog", the transformer should be mapping each token to an embedding in the end (attending each to each). I want the embedding that the last attention layer assigns to the last token. Not the logits themselves, but certainly not just the word embeddings from the beginning -before any feedforward through the attention layers-.

StrikingLoo committed:
Working! Thanks to @nullhook

nullhook commented:
The input embeddings obtained at the starting layer are static and not contextual? In order for the embeddings to include context for each input token in the sequence, it is necessary for the input to go through all the layers?

StrikingLoo committed:
merge + working embeddings

StrikingLoo commented:
Sorry, I don't get if those are assertions or questions. I understand we want for the whole input to go through all the attention layers, then we take the representation for the last token. That is what this PR is supposed to be doing. Right now I think it is working, though I would say we can test it more before merging (right now I see it doesn't break, and it gives an output of the correct shape, but I would test more)

StrikingLoo commented:
Okay, I think this is working well. I tested it with multiple inputs and the embeddings make a lot of sense.
Just to clarify, what this is doing is:
- Take the input (a sentence, etc.)
- Feed it through all the attention layers. In the end we have N embeddings of size `n_embd`.
- Print the embedding corresponding to the last token in the input.
- Stop execution
Right now it works without errors both in embedding mode and not embedding mode. The embeddings look coherent. More exhaustive testing may be performed with someone with access to better compute, but I tried 8 inputs, some semantically related, some not, and the cosine similarities between embeddings made sense. I made some sentences have the exact same last word so that, if we were merely outputting word embeddings with no feedforwarding, similarity should be maximum, which it was not, so this looks good.

Do let me know if you think other changes are needed before merging, and thank you for your help in getting this to work!

nullhook commented on main.cpp:
```diff
@@ -1089,12 +1114,28 @@ int main(int argc, char ** argv) {
+        embd = embd_inp;
+        if (embd.size() > 0) {
+            const int64_t t_start_us = ggml_time_us(); 
+            if (!llama_eval(model, params.n_threads, n_past, embd, logits, mem_per_token, true)) {
```
pass in the `params.embedding` instead, which presumably contains the same information about whether or not to use the embeddings as input

nullhook commented on main.cpp:
```diff
@@ -1089,12 +1114,28 @@ int main(int argc, char ** argv) {
+    if (params.embedding){
+        embd = embd_inp;
+        if (embd.size() > 0) {
+            const int64_t t_start_us = ggml_time_us(); 
```
It seems that we are storing the start timestamp here, but is it being used to measure anything in embedding mode? If not, let's remove it. Alternatively, do we want to measure the embedding compute time?

StrikingLoo committed:
make params argument instead of hardcoded boolean. remove useless time check

StrikingLoo committed:
merge with main, move logic for embeddings into llama.cpp

StrikingLoo commented:
Okay, merged with master again, moved everything to llama.cpp. I addressed both changes: now everywhere that makes sense, the boolean matches the console arguments (the test in the beginning for memory size needs to be hardcoded to false) and I removed the spurious time check.

ggerganov reviewed (changes requested):
Hi, this will be a great addition!

Let's change slightly the interface like this:

- Add parameter to `llama_context_params` called `embeddings` which is `false` by default. It will be used to specify if we want the `llama_eval()` call to keep the computed embeddings or not
- Add an embedding buffer (`std::vector<float>`) to the `llama_context` similar to the `logits` buffer and resize it upon model load if the `embeddings` parameter is `true`. If  it is `false`, then leave it empty
- Keep `llama_eval_internal()` as original and only change to copy the embeddings into the new context buffer if it is not empty (i.e. `embeddings` parameter was set to `true` during init). You will need to store the `ggml` tensor at the end in a separate variable so you can access it after the `ggml_graph_compute()` call, similar to the `logits`. No need for an alternative `ggml_graph_compute()` call as you have proposed
- Add C API call to get the new embeddings buffer - same as `llama_get_logits()`
- You can now print the embeddings in user code, instead of in the `llama` library

nullhook commented:
I'm curious if normalizing reduces the dimensionality of the embedding space? If the goal of the consumer is to only compute cosine similarity from the embeddings, then I think it makes sense to normalize. However, I suggest leaving the embeddings as raw vectors and letting the consumer decide whether or not to normalize them. It's possible that others may have a different opinion on this.

nullhook commented:
It would be helpful to add an example of generating input embeddings, normalizing them, and computing cosine similarity in the example folder with this PR.

StrikingLoo committed:
merge

StrikingLoo commented:
Hi ggerganov, 
Thank you for the instructions. I did steps 1, 2 and 4, but I am not sure what we want the final program to look like in step 3.

"Keep llama_eval_internal() as original and only change to copy the embeddings into the new context buffer if it is not empty (i.e. embeddings parameter was set to true during init). You will need to store the ggml tensor at the end in a separate variable so you can access it after the ggml_graph_compute() call, similar to the logits. No need for an alternative ggml_graph_compute() call as you have proposed"

Do I still keep the behavior where we check and, if embeddings==true, we only run until we have the embeddings then return? If I do I can't leave llama_eval_internal identical. Plus there is that test run we do at the beginning (Something about memory cost) that requires me to send a false for the embeddings, even if we will keep them later.

The model load / config is already as specified.

StrikingLoo committed:
start doing the instructions but not finished. This probably doesnt compile

ggerganov committed:
Embeddings extraction support

ggerganov commented:
@StrikingLoo Just pushed a change. Give it a try and let me know if it works

StrikingLoo commented:
This looks perfect to me. Just to make sure: would we be merging like this, or should I add the display_embeddings code under "// TODO: print / use the embeddings"?

ggerganov commented:
This will become a separate example program called `embedding`. For now will merge like this because I want to start refactoring and later we will move into a separate program

loretoparisi commented:
@ggerganov is this the correct command
```
./embedding -m models/7B/ggml-model-q4_0.bin -p "ciao" -n 512 
```

It seems it's not using the prompt in `p`.  Infact I do not see in the logs the log I see in `embedding.cpp` that should print it:

```
main: seed = 1679853514
llama_model_load: loading model from 'models/7B/ggml-model-q4_0.bin' - please wait ...
llama_model_load: n_vocab = 32000
llama_model_load: n_ctx   = 512
llama_model_load: n_embd  = 4096
llama_model_load: n_mult  = 256
llama_model_load: n_head  = 32
llama_model_load: n_layer = 32
llama_model_load: n_rot   = 128
llama_model_load: f16     = 2
llama_model_load: n_ff    = 11008
llama_model_load: n_parts = 1
llama_model_load: type    = 1
llama_model_load: ggml ctx size = 4273.34 MB
llama_model_load: mem required  = 6065.34 MB (+ 1026.00 MB per state)
llama_model_load: loading model part 1/1 from 'models/7B/ggml-model-q4_0.bin'
llama_model_load: .................................... done
llama_model_load: model size =  4017.27 MB / num tensors = 291
llama_init_from_file: kv self size  =  256.00 MB

system_info: n_threads = 4 / 8 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
2.177604 -1.095253
...
04 0.534440 0.732717 0.781988 -1.836264 -0.860989 -0.564879 0.084990 0.838598 1.210304 -0.441369 -1.963783 2.096257 

llama_print_timings:        load time =  1124.05 ms
llama_print_timings:      sample time =     0.00 ms /     1 runs   (    0.00 ms per run)
llama_print_timings: prompt eval time =   464.57 ms /     3 tokens (  154.86 ms per token)
llama_print_timings:        eval time =     0.00 ms /     1 runs   (    0.00 ms per run)
llama_print_timings:       total time =  2215.86 ms
```
